x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/entsearch/engine/action/PutEngineAction.java

```diff
@@ -0,0 +1,129 @@
+    public static final String NAME = "indices:admin/engine/put";
+
+    public PutEngineAction() {
+        super(NAME, PutEngineAction.Response::new);
+    }
+
+    public static class Request extends ActionRequest implements IndicesRequest.Replaceable {
+
+        private final Engine engine;
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            this.engine = new Engine(in);
+        }
+
+        public Request(String engineId, BytesReference content, XContentType contentType) {
+            this.engine = Engine.fromXContentBytes(engineId, content, contentType);
+        }
+
+        public Request(Engine engine) {
+            this.engine = engine;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            ActionRequestValidationException validationException = null;
+
+            if (engine.indices().length == 0) {
+                validationException = addValidationError("indices are missing", validationException);
+            }
+
+            return validationException;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            engine.writeTo(out);
+        }
+
+        @Override
+        public String[] indices() {
+            return engine.indices();
+        }
+
+        @Override
+        public IndicesRequest indices(String... indices) {
+            return new Request(new Engine(engine.name(), indices, engine.analyticsCollectionName()));
+        }
+
+        @Override
+        public IndicesOptions indicesOptions() {
+            return IndicesOptions.fromOptions(false, false, false, false, false, false, false, false);
+        }
+
+        public Engine getEngine() {
+            return engine;
+        }
+    }
+
+    public static class Response extends ActionResponse implements StatusToXContentObject {
+
+        final DocWriteResponse.Result result;
+
+        public Response(StreamInput in) throws IOException {
+            super(in);
+            result = DocWriteResponse.Result.readFrom(in);
+        }
+
+        public Response(DocWriteResponse.Result result) {
+            this.result = result;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            this.result.writeTo(out);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field("result", this.result.getLowercase());
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public RestStatus status() {
+            return switch (result) {
+                case CREATED -> RestStatus.CREATED;
+                case NOT_FOUND -> RestStatus.NOT_FOUND;
+                default -> RestStatus.OK;
```

Comment:
Why do we need to restrict the status code? It should be possible to use this action to override an existing Engine, no?

Reply:
Yes, in that case the status will be OK - there's no specific status for an update AFAIU

Reply:
Ok cool, thanks

Reply:
Why do we need to return a 404 for the PUT command?

Reply:
It depends on what the response issued from us would be. I just didn't want a 404 result to pass through undetected - for example, it could happen in the alias response in case it is not handled properly for changing that into a 400.